### PR TITLE
filedelete: re-write plugin to clarify model

### PR DESCRIPTION
**Pull request by skvl**
Before refactoring multiple hooks on NtClose been set for each
step in injection chain. Thus the model been difficult to understand.
The other drawback was that the order of callbacks been called was
crucial.

With new model the single hook on NtClose left. And every task get
it's state machine. Thus the model been clarified a lot.

### CODE REVIEW

**Comment by tklengyel** [comment]
Please rebase on latest master

**Comment by skvl** [comment]
Sorry. I didn't noticed that I failed to fetch new master.

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by skvl** [comment]
@tklengyel , is something wrong with PR?

File: src/plugins/filedelete/filedelete.cpp [hunk 1]
```diff
@@ -122,1348 +121,1514 @@
...
+                &filedelete::createsection_cb);
+        this->createfile_hook = createSyscallHook("NtCreateFile",
+                &filedelete::createfile_cb);
+        this->openfile_hook = createSyscallHook("NtOpenFile",
+                &filedelete::openfile_cb);
+    }
 }
 
-static bool get_file_object_flags(drakvuf_t drakvuf, drakvuf_trap_info_t* info, vmi_instance_t vmi, filedelete* f, handle_t handle, uint64_t* flags)
+/* NOTE One should rund drakvuf loop to restore VM state.
```
Comment (tklengyel) [inline src/plugins/filedelete/filedelete.cpp:201]: `rund`

File: src/plugins/filedelete/filedelete.cpp [hunk 2]
```diff
@@ -122,1348 +121,1514 @@
...
+ *
+ * The plug-in injects syscalls thuns changes the state. So to avoid BSOD
```
Comment (tklengyel) [inline src/plugins/filedelete/filedelete.cpp:203]: `thuns`

File: src/plugins/filedelete/filedelete.cpp [hunk 3]
```diff
@@ -122,1348 +121,1514 @@
...
+ * one should restore state here. This requires to allow VM to live for a
+ * while.
+ *
+ * Hint: there is no need to waite all files read finish. Just waite every
```
Comment (tklengyel) [inline src/plugins/filedelete/filedelete.cpp:207]: `waite`

**Comment by tklengyel** [comment]
So overall this looks fine to me but the number of changes to the filedelete plugin is so significant that it's hard to follow. Since this is almost a reimplementation of the plugin I would suggest actually just placing it into a new plugin (`filedelete2`) and defaulting to this, but adding a configure override to allow using the pre-existing version in case there are discrepancies.

**Comment by skvl** [comment]
@tklengyel , I have rename plug-in to "fileextractor". I think it is ready.

**Comment by skvl** [comment]
@tklengyel , I don't understand the errors from CodeQL: https://github.com/tklengyel/drakvuf/pull/1393/checks?check_run_id=5245213598

File: src/libhook/call_result.hpp [hunk 4]
```diff
@@ -118,11 +118,13 @@ struct CallResult
...
     virtual ~CallResult() = default;
 
-    void setResultCallParams(const drakvuf_trap_info_t* info)
+    void setResultCallParams(const drakvuf_trap_info_t* info, void* context_ = nullptr)
     {
         target_pid = info->attached_proc_data.pid;
         target_tid = info->attached_proc_data.tid;
         target_rsp = info->regs->rsp;
+        context = context_;
+
```
Comment (BonusPlay) [inline src/libhook/call_result.hpp:127]: Please dont do this. Instead create a new structure inheriting from PluginResult (https://github.com/tklengyel/drakvuf/blob/master/src/plugins/hook_helpers.h#L128-L136) or store the data on plugin (`this`) object.

File: src/plugins/fileextractor/fileextractor.cpp [hunk 5]
```diff
@@ -0,0 +1,1634 @@
...
+            this->dump_folder,
+            task.idx) < 0 )
+        return;
+
+    FILE* fp = fopen(file, "w");
+    free(file);
+    if (!fp)
+        return;
+
+    json_object* jobj = json_object_new_object();
```
Comment (BonusPlay) [inline src/plugins/fileextractor/fileextractor.cpp:1156]: maybe use fmt instead of raw json?

**Review by BonusPlay — CHANGES_REQUESTED** [review]

**Comment by skvl** [comment]
> CodeQL just complains about the default file permissions when a new file is created. Just add `umask(S_IWGRP|S_IWOTH);` before the `fopen`.

Fixed

File: src/libhook/call_result.hpp (hunk 4 above)
Comment (skvl) [reply src/libhook/call_result.hpp]: Fixed

File: src/plugins/fileextractor/fileextractor.cpp (hunk 5 above)
Comment (skvl) [reply src/plugins/fileextractor/fileextractor.cpp:1156]: Why should we do this?

The usage of `fmt` would be against the switch to json-c library by @icedevml from https://github.com/tklengyel/drakvuf/commit/048e9e343a800ecabaeeb8f958c6341c1b837f97

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please
